# Organization-and-annotation-of-eukaryote-genomes

## Description and goal of the project

This repository contains scripts for the annotation of an assembly of *Arabidopsis thaliana* accession Lu-1. The assembly of this accession was generated during the previous course of Genome and Transcriptome assembly of the Master's in Bioinformatics and Computational Biology. More information about the assembly can be found [here](https://github.com/valentin-muller26/Genome-and-Transcriptome-Assembly)
## Datasets

**Assembly Data:**
The assembly for the genomic data comes from the accession Lu-1 and was sequenced using PacBio HiFi reads and assembled using Hifiasm version 0.25.0.

**Transcriptomic Data:**
The assembly for the transcriptomic data comes from the accession Sha and was sequenced using Illumina paired-end short reads and assembled using Trinity version 2.15.1.

The sequencing data can be found in the following links:  
- Genome data: https://www.nature.com/articles/s41588-024-01715-9
- Transcriptome data: https://www.nature.com/articles/s41467-020-14779-y

## TE Annotation Pipeline

### 1.a Transposable element (TE) Annotation using EDTA `01a_run_EDTA.sh`

In this first step we used EDTA (Extensive de novo TE Annotator) version 2.2 to annotate the TEs in our assembly using the script `01a_run_EDTA.sh`. This annotation resulted in a folder containing TE library for the whole genome and classification of the elements into superfamilies.

More information about EDTA can be found [here](https://github.com/oushujun/EDTA)

### 1.b-d Analysing the Full-length Long Terminal Repeats (LTR) Retrotransposons 

Once EDTA has generated the annotation of the retrotransposons, we perform an analysis of the intact full-length LTR retrotransposons to have an insight of their percent identity. This percent identity is a measure of the similarity between the sequence of the two flanking LTR of the retrotransposon, this measure help us to infer the relative age of the element and the time since insertion. We can infer those time and age because when a Retrotransposon is inserted the two LTR are similar but with time the LTR accumulate mutation thus the percent identity decreases between them.

This analysis was performed in three steps:

**Step 1: Extract LTR percent identity `01b_extract_LTR_identity.sh`**

This script parses the main information of intact full-length LTR retrotransposons and provides some summary statistics.

**Step 2: Refine LTR-RT classification with TEsorter `01c_run_TEsorter_FullLTR_classification.sh`**

This step uses TEsorter version 1.3.0 to refine the classification of the intact LTRs into clades.

**Step 3: Visualize LTR-RT dynamics `01d-full_length_LTRs_identity.R`**
This step uses R version 4.5.0 to generate histograms showing the distribution of LTR percent identity for each clade within the Copia and Gypsy superfamilies.

This step of the analysis is done locally and the files `HiFiasm_Lu1_primary.fa.mod.EDTA.intact.raw.gff3`, generated by EDTA, and `HiFiasm_Lu1_primary.fa.mod.LTR.raw.fa.rexdb-plant.cls.tsv`, generated by TE Sorter, need to be downloaded from the cluster and placed in the same directory as the `01d-full_length_LTRs_identity.R` script.

## 2. Visualizing and comparing TE annotations from EDTA
In this step we use the results generated by EDTA, and we visualize and compare them to gain insights into the transposable element landscape of the accession.

### 2.a Generate FAI index file `02a_generate_fai.sh`

To prepare for visualization, we first generate a FAI index file from the assembly using samtools version 1.13-GCC-10.3.0.


### 2.b Visualize TE density across the genome `02b-annotation_circlize.R`
This step uses R version 4.5.0 with packages `circlize` to generate circular plots showing the distribution and density of different TE superfamilies across the genome.

This analysis is done locally and the following input files need to be downloaded:
- `HiFiasm_Lu1_primary.fa.mod.EDTA.TEanno.gff3`: TE annotation from EDTA
- `HiFiasm_Lu1_primary.fa.fai`: Genome index file (generated in step 2.a)
- `Copia.cls.tsv`: Copia clade classification from TEsorter (generated in step 3) 
- `Gypsy.cls.tsv`: Gypsy clade classification from TEsorter (generated in step 3)

The script generates the following visualization files :
**Circos plots:**
- `02-TE_density.pdf`: Comparison of Gypsy and Copia LTR retrotransposon densities
- `02-TE_density_all_superfamilies.pdf`: Density of the top 6 most abundant TE superfamilies
- `03-Athila_and_CRM.pdf`: Distribution of Athila and CRM clades (centromeric TEs in Brassicaceae)

**Bar plots:**
- `06-Copia_Gypsy_clades_abundance.pdf`: Side-by-side comparison of Copia and Gypsy clade abundances
- `07-Copia_Gypsy_Complete_Incomplete.pdf`: Stacked bar plot showing complete vs incomplete elements for each clade


## 3. Refining TE Classification
The goal of this step is to refine the classification of the Copia and Gypsy superfamilies to have a clade resolution.
The refining is done in two step first the script `03a_extract_Copia_Gypsy.sh` was used to extract the Copia and the Gypsy superfamilies retrotransposon from the `.mod.EDTA.TElib.fa` and generate for each superfamilies a fasta file. Those fasta files were used with TEsorter (script `03b_run_TEsorter_Copia_Gypsy.sh`) to refine and have a clade resolution of those retrotransposon.

## 4. Dynamics of Transposable Elements (TEs)
The goal of this step is to estimate the insertion age of TEs by calculating the divergence of their sequences from consensus sequences to understand the evolutionary dynamics of transposable elements in the genome.

This analysis is done using two scripts :

### 4.a Parse RepeatMasker outputs `04a_parseRM.sh`

This script uses `parseRM.pl` to process the RepeatMasker alignment outputs (`genomic.fna.mod.out`) generated by EDTA in step 1. The script calculates the corrected percentage of divergence for each TE copy from its consensus sequence, accounting for the high mutation rate at CpG sites.

- Low divergence values indicate recent TE insertions
- High divergence values suggest older insertions

### 4.b Visualize TE landscape `04b_plot_div.R`

This step uses R version 4.5.0 to generate TE landscape plots showing the divergence distribution of TEs across the genome and their estimated insertion age.

This analysis is done locally and the following input file needs to be downloaded:
- `HiFiasm_Lu1_primary.fa.mod.out.landscape.Div.Rname.tab`: Divergence data from parseRM (generated in step 4.a)

The script generates two visualization files:
**Visualization (saved in `plots/` directory)**
- `TE_landscape_distance.pdf`: TE landscape showing divergence from consensus sequences
- `TE_landscape_age.pdf`: TE landscape showing estimated insertion age in million years

## 5. Gene Annotation with MAKER
The goal of this step is to generate gene prediction using MAKER by performing evidence-based gene prediction that integrates transcriptomic evidence, protein homology, and ab initio predictions.

### 5.a Generate and configure MAKER control files `05a_run_maker_setup.sh`

This script generates the four control files required by MAKER and automatically configures the main parameters in `maker_opts.ctl`.

The control files generated are:
- `maker_opts.ctl`: Main configuration file with all annotation parameters
- `maker_bopts.ctl`: BLAST and filtering options
- `maker_evm.ctl`: Evidence modeler weights
- `maker_exe.ctl`: Paths to external executables

**Parameters modified in maker_opts.ctl:**
- `genome`: HiFiasm_Lu1_primary.fa
- `est`: Trinity transcriptome assembly from Sha accession
- `protein`: TAIR10 representative gene models and UniProt plant reviewed proteins
- `model_org`: disabled (empty value to save computational time)
- `rmlib`: HiFiasm_Lu1_primary.fa.mod.EDTA.TElib.fa (TE library from EDTA)
- `repeat_protein`: PTREP20 database
- `augustus_species`: arabidopsis
- `est2genome`: 1 (enabled)
- `protein2genome`: 1 (enabled)
- `cpus`: 1 (MPI handles parallelization)
- `alt_splice`: 1 (enabled)
- `TMP`: Path for temporary files storage

### 5.b Run MAKER with MPI `05b_run_maker.sh`
The annotation is performed using MAKER version 3.01.03 with MPI for parallel processing and requires the modules OpenMPI version 4.1.1 and AUGUSTUS version 3.4.0. The annotation combines evidence from transcriptomic and protein alignments, repeat masking using the EDTA TE library, and predictions using Augustus.

### 5.c Merge the result of the MAKER gene annotation `05c_maker_merge_result.sh`
The result of the MAKER gene annotation is a datastore containing the annotation in separate directories and output files for each region (contigs/scaffolds). The script `05c_maker_merge_result.sh` will merge the different results in the following outputs:

**Gff files**
- `HiFiasm_Lu1_primary.all.maker.gff`: combining all gene predictions with their sequences
- `HiFiasm_Lu1_primary.all.maker.noseq.gff`: combining all gene predictions without their sequences

**FASTA files**
- `HiFiasm_Lu1_primary.all.maker.proteins.fasta`: containing all the predicted protein sequences
- `HiFiasm_Lu1_primary.all.maker.transcripts.fasta`: containing all predicted transcript sequences
- `HiFiasm_Lu1_primary.all.maker.augustus_masked.proteins.fasta`: containing the list of proteins from Augustus predictions
- `HiFiasm_Lu1_primary.all.maker.augustus_masked.transcripts.fasta`: containing the list of transcripts from Augustus predictions
- `HiFiasm_Lu1_primary.all.maker.non_overlapping_ab_initio.proteins.fasta`: containing the list of unique ab initio protein predictions (proteins not supported by other evidence than ab initio)
- `HiFiasm_Lu1_primary.all.maker.non_overlapping_ab_initio.transcripts.fasta`: containing the list of unique ab initio transcript predictions (transcripts not supported by other evidence than ab initio)

### 5.d Filter and refine MAKER gene annotations `05d_filter_gene_annotation.sh`

After merging the MAKER outputs, the annotations were filtered to retain only high-quality gene models. This filtering is performed in the following steps:

**Step 1: Assign consistent gene IDs**
In this first step we used MAKER's ID mapping tool to replace the IDs in the GFF and FASTA files generated by MAKER with cleaner and more consistent IDs.

**Step 2: Functional annotation with InterProScan**
In this step we used InterProScan on the predicted protein sequences to identify and annotate protein domains, families, and Gene Ontology (GO) terms using the Pfam database. These functional annotations from InterProScan were integrated into the GFF file.

**Step 3: Calculate Annotation Edit Distance (AED) values**
We used MAKER's AED_cdf_generator.pl to generate AED values. These values measure the agreement between gene predictions and the evidence used (transcripts, proteins, ab initio predictions). Lower AED values indicate better support.

**Step 4: Filtering the GFF**
The GFF files were first filtered using MAKER's quality filter tool to retain only gene models with high evidence support. Then they were filtered to retain only the following gene features (gene, CDS, exon, 5'UTR, 3'UTR, mRNA).

**Step 5: Extract filtered sequences** 
The corresponding protein and transcript sequences were extracted based on the filtered gene list generated in the previous step.

**Final outputs:**
- `filtered.genes.renamed.gff3`: Filtered gene annotations with functional information
- `HiFiasm_Lu1_primary.all.maker.proteins.renamed.filtered.fasta`: Filtered protein sequences
- `HiFiasm_Lu1_primary.all.maker.transcripts.renamed.filtered.fasta`: Filtered transcript sequences

### 5.e Visualize TE and density across the genome `05e_annotation_circlize_gene.R`
This script generates the same Circos plots as in step 2.b with the addition of a track for gene density. The gene annotations from the filtered MAKER (`filtered.genes.renamed.gff3`) output are used to display the distribution of genes alongside TE families and superfamilies.

### 5.f (Optional) Generating AED score distribution plots `05f_generate_AED_plot.R`
This script generates visualizations of the Annotation Edit Distance (AED) score distribution from the MAKER annotation. It requires the AED analysis output file (`assembly.all.maker.renamed.gff.AED.txt`) generated in step 5.d.

**Outputs:**
- `07-AED_distribution_histogram.pdf`: Histogram showing the distribution of AED scores across all predicted genes
- `07-AED_cumulative_distribution.pdf`: Cumulative distribution plot of AED scores with threshold indicators at 0.25 and 0.5

## 6. Quality Assessment of Gene Annotations
After filtering the MAKER annotations, the quality of the gene predictions was performed using BUSCO (Benchmarking Universal Single-Copy Orthologs) and AGAT :

### 6.a Extracting longest isoforms 
Before running the quality assesment the longest isoform of each gene was extracted using the script `06a_extract_longest.sh`

This script extracts:
- The longest protein sequence per gene from `HiFiasm_Lu1_primary.all.maker.proteins.renamed.filtered.fasta`
- The longest transcript sequence per gene from `HiFiasm_Lu1_primary.all.maker.transcripts.renamed.filtered.fasta`

### 6.b Running BUSCO `06b_run_Busco.sh`
BUSCO version 5.4.2 is run on both the longest protein and transcript sequences and the `brassicales_odb10` lineage dataset to assess the completeness of our annotation.

### 6.c (Optional) Generating plot for the result of Busco `06c_busco_plot.sh`
This script uses BUSCO version 5.4.2 to generate individual and combined plot consisting of the result of the busco analysis of the annotation for protein and transcript sequences and the genomic and transcriptomic assembly.

### 6.d Running AGAT: Annotation Statistics `06d_run_agat.sh`
In addition to BUSCO, AGAT version 1.5.1 was run on our gene annotation to generate comprehensive statistics.

## 7 Functional Annotation with Protein Homology `07a_run_homology.sh`
This script performs protein homology searches to retrieve functional annotations for the genes predicted and filtered in previous steps using two reference databases.

The first search was performed against UniProt, where the predicted proteins are compared against the UniProt reviewed plant proteins database (uniprot_viridiplantae_reviewed.fa) using BLASTP with an e-value threshold of 1e-5. The resulting matches were sorted and only the best hit per query was kept. The functional annotations retrieved from these best hits were mapped to the GFF3 and FASTA files generated by MAKER. Additionally, a comparison against TAIR10 (reference model for *Arabidopsis thaliana*) was performed following the same steps as for the UniProt comparison.

**Main Outputs:**
- `blastp_uniprot.out.besthits`: Best UniProt hit per protein
- `blastp_tair10.out.besthits`: Best TAIR10 hit per protein
- `filtered.genes.renamed.gff3.Uniprot.annotated`: GFF3 with functional annotations
- `HiFiasm_Lu1_primary.all.maker.proteins.fasta.Uniprot.annotated`: Proteins with functional annotations

## 8 Pangenome analysis using Genespace
In this step we used GeneSpace to conduct a small pangenome analysis comparing the Lu-1 accession against TAIR10,Altai-5,Are-6 and Etna-2 accession. This analyis was perform in the following step : 

**Step 1 : Setting the files for the Genespace analysis `08a_setup_genespace.sh`**

This script prepares the required input files for GENESPACE analysis by generating BED files and FASTA files with matching gene names in the BED file for the five accession.

**Step 2 : Running GENESPACE `08b_genespace.R` and `08c_run_genespace.sh`**

The bash script executes the R script to run the GENESPACE analysis, which performs synteny and orthology inference. The analysis generates a pangenome matrix showing orthogroup assignments across all accessions.

**Main Outputs:**
- pangenome_matrix.rds : matrix showing orthogroup assignments across all accessions
- Riparian plot showing syntenic relationships between accessions
- Dotplot : plot showing syntenic relationships between pairwise accession comparison

**Step 3 : Generating visualizations of the orthogroups and Core, Accessory and Unique Gene `08d_process_pangenome.R`**
This script processes the pangenome matrix generated by GENESPACE to analyze and visualize the distribution of orthogroups and genes across accessions. The analysis categorizes orthogroups into three groups based on their presence:

- **Core:** Present in all 5 accessions
- **Accessory:** Present in 2-4 accessions
- **Unique (species-specific):** Present in only one accession

This script generate statistic of number of orthogroup and genes in the different category for each accession and for all accession and generate the following visualizations :
- `pangenome_frequency_plot.pdf`: Bar plot showing the distribution of orthogroups and genes by the number of genomes they are present in
- `genes_plot.pdf`: Stacked bar plot showing the number of core, accessory and unique genes per accession

## List of the tools used

| Tool | Version | 
|------|---------|
| EDTA | 2.2 | 
| TEsorter | 1.3.0 |
|SAMtools |1.13|
| SeqKit | 2.6.1 | 
| parseRM.pl | - | 
| circlize | R package |
| MAKER | 3.01.03 | 
| OpenMPI | 4.1.1 |
| BioPerl | 1.7.8 |
|AUGUSTUS| 3.4.0|
| R | 4.5.0 |
| InterProScan | 5.70-102.0|
| UCSC-Utils | 448 |
| MariaDB | 10.6.4 | 
| BUSCO | 5.4.2 |
| AGAT | 1.5.1 |
| BLAST+ | 2.15.0 |
| GENESPACE | -|


## Author

Valentin MÃ¼ller
