# Organization-and-annotation-of-eukaryote-genomes

## Description and goal of the project

This repository contains scripts for the annotation of an assembly of *Arabidopsis thaliana* accession Lu-1. This assembly was done during the genome and transcriptome assembly course. 

## Datasets

**Assembly Data:**
The assembly for the genomic data comes from the accession Lu-1 and was sequenced using PacBio HiFi reads and assembled using Hifiasm version 0.25.0.

**Transcriptomic Data:**
The assembly for the transcriptomic data comes from the accession Sha and was sequenced using Illumina paired-end short reads and assembled using Trinity version 2.15.1.

The sequencing data can be found in the following links:  
- Genome data: https://www.nature.com/articles/s41588-024-01715-9
- Transcriptome data: https://www.nature.com/articles/s41467-020-14779-y

## TE Annotation Pipeline

### 1.a TE Annotation using EDTA `01a_run_EDTA.sh`

In this first step we used EDTA (Extensive de novo TE Annotator) version 2.2 to annotate the TEs in our assembly using the script `01a_run_EDTA.sh`. This annotation resulted in a folder containing TE library for the whole genome and classification of the elements into superfamilies.

More information about EDTA can be found [here](https://github.com/oushujun/EDTA)

### 1.b-d Analysing the Full-length LTR Retrotransposons 

Once EDTA has generated the annotation of the retrotransposons, we perform an analysis of the intact full-length LTR retrotransposons to have an insight of their percent identity. This percent identity is a measure of the similarity between two DNA sequences and for the LTR of retrotransposon, this measure help us to infer the relative age of the element and the time since insertion. We can infer those time and age because when a Retrotransposon is inserted the two LTR are similar but with time the LTR accumulate mutation this the percent identity decreases between them.

This analysis was performed in three steps:

**Step 1: Extract LTR percent identity `01b_extract_LTR_identity.sh`**

This script parses the main information of intact full-length LTR retrotransposons and provides some summary statistics.

**Step 2: Refine LTR-RT classification with TEsorter `01c_Run_TEsorter_FullLTR_classification.sh`**

This step uses TEsorter version 1.3.0 to refine the classification of the intact LTRs into clades.

**Step 3: Visualize LTR-RT dynamics `01d-full_length_LTRs_identity.R`**
This step uses R version 4.5.0 to generate histograms showing the distribution of LTR percent identity for each clade within the Copia and Gypsy superfamilies.

This analysis is done locally and the files `genomic.fna.mod.LTR.intact.raw.gff3` and `genomic.fna.mod.LTR.raw.fa.rexdb-plant.cls.tsv` need to be downloaded from the cluster and placed in the same directory as the `01d-full_length_LTRs_identity.R` script.

## 2. Visualizing and comparing TE annotations from EDTA
In this step we use the results generated by EDTA, and we visualize and compare them to gain insights into the transposable element landscape of the accession.

### 2.a Generate FAI index file `02a_generate_fai.sh`

To prepare for visualization, we first generate a FAI index file from the assembly using samtools version 1.13-GCC-10.3.0.


### 2.b Visualize TE density across the genome `02b-annotation_circlize.R`
This step uses R version 4.5.0 with packages `circlize` to generate circular plots showing the distribution and density of different TE superfamilies across the genome.

This analysis is done locally and the following input files need to be downloaded:
- `HiFiasm_Lu1_primary.fa.mod.EDTA.TEanno.gff3`: TE annotation from EDTA
- `HiFiasm_Lu1_primary.fa.fai`: Genome index file (generated in step 2.a)
- `Copia.cls.tsv`: Copia clade classification from TEsorter (generated in step 3.b) 
- `Gypsy.cls.tsv`: Gypsy clade classification from TEsorter (generated in step 3.b)

The script generates five visualization files and summary files:
**Visualization (saved in `plots/` directory)**
- `02-TE_density.pdf`: Comparison of Gypsy and Copia LTR retrotransposon densities
- `02-TE_density_all_superfamilies.pdf`: Density of the top 6 most abundant TE superfamilies
- `03-Athila_and_CRM.pdf`: Distribution of Athila and CRM clades (centromeric TEs in Brassicaceae)
- `04-Clade_abundance_barplot.pdf`: Bar plot of the top 15 most abundant TE clades
- `05-Copia_vs_Gypsy_clades.pdf`: Side-by-side comparison of Copia and Gypsy clade distributions
**Summary files saved in `results/` directory**
- `Copia_clades_summary.tsv`: Statistics for all Copia clades (count, complete/incomplete elements, percentage)
- `Gypsy_clades_summary.tsv`: Statistics for all Gypsy clades (count, complete/incomplete elements, percentage)
- `Overall_clades_summary.tsv`: Overall summary of all TE clades identified
- `TE_clades_comprehensive_report.xlsx`: Excel file with all summaries


## 3. Refining TE Classification
The goal of this step is to refine the classification of the Copia and Gypsy superfamilies to have a clade resolution.
The refining is done in two step first the script `03a_extract_Copia_Gypsy.sh` was used to extract the Copia and the Gypsy superfamilies retrotransposon from the `.mod.EDTA.TElib.fa` and generate for each superfamilies a fasta file. Those fasta files whose used with TeSorter (script `03b_run_TEsorter_Copia_Gypsy.sh`) to refine and have a clade resolution of those retrotransposon.

## 4. Dynamics of Transposable Elements (TEs)
The goal of this step is to estimate the insertion age of TEs by calculating the divergence of their sequences from consensus sequences to understand the evolutionary dynamics of transposable elements in the genome.

This analysis is done using two scripts :

### 4.a Parse RepeatMasker outputs `04a_parseRM.sh`

This script uses `parseRM.pl` to process the RepeatMasker alignment outputs (`genomic.fna.mod.out`) generated by EDTA in step 1. The script calculates the corrected percentage of divergence for each TE copy from its consensus sequence, accounting for the high mutation rate at CpG sites.

- Low divergence values indicate recent TE insertions
- High divergence values suggest older insertions

### 4.b Visualize TE landscape `04b_plot_div.R`

This step uses R version 4.5.0 to generate TE landscape plots showing the divergence distribution of TEs across the genome and their estimated insertion age.

This analysis is done locally and the following input file needs to be downloaded:
- `HiFiasm_Lu1_primary.fa.mod.out.landscape.Div.Rname.tab`: Divergence data from parseRM (generated in step 4.a)

The script generates two visualization files:
**Visualization (saved in `plots/` directory)**
- `TE_landscape_distance.pdf`: TE landscape showing divergence from consensus sequences
- `TE_landscape_age.pdf`: TE landscape showing estimated insertion age in million years

## List of the tools used

| Tool | Version | 
|------|---------|
| EDTA | 2.2 | 
| TEsorter | 1.3.0 |
|SAMtools |1.13|
| SeqKit | 2.6.1 | 
| parseRM.pl | - | 
| circlize | R package |
| MAKER | 3.01.03 | 
| OpenMPI | 4.1.1 |
| BioPerl | 1.7.8 |
| R | 4.5.0 |


## Author

Valentin MÃ¼ller